mailpoet/lib/Subscribers/InactiveSubscribersController.php

```diff
@@ -98,7 +96,7 @@ private function deactivateSubscribers(Carbon $thresholdDate, int $batchSize, ?i
       $this->inactiveTaskIdsTableCreated = true;
     }
 
-    // Select subscribers who received at least a number of tracked emails but didn't open any
+    // Select subscribers who received at least a number of emails after  threshold date
```

Comment:
Another comment that was previously pretty confusing. It said `Select subscribers who received a recent tracked email but didn't open it but there is no check for "not opening it" as it only selects subscribers with a "recently received emails but subscribed non-recently".

Could we maybe improve the comment to sth like `Select subscribers who received at least a number of emails after threshold date and subscribed before that`?

Reply:
I also really wonder why the part with checking engagement is not directly here — why there is yet another query below. Again, this is not a topic for this PR but I think we would save ourselves a query, a temporary table, and loading a lot of IDs to memory by just doing it all here:

```sql
SELECT s.id FROM {$subscribersTable} as s
  JOIN {$scheduledTaskSubscribersTable} as sts USE INDEX (subscriber_id) ON s.id = sts.subscriber_id
  JOIN {$inactiveTaskIdsTable} task_ids ON task_ids.id = sts.task_id
WHERE s.last_subscribed_at < :thresholdDate
  AND s.status = :status
  AND s.id >= :startId
  AND s.id < :endId
  AND s.last_engagement_at > :thresholdDate -- I added this
GROUP BY s.id
HAVING count(s.id) >= :unopenedEmailsThreshold
```

Note that I added only `AND s.last_engagement_at > :thresholdDate` and not the other columns as they seem to make no sense:
```sql
AND GREATEST(
  COALESCE(s.last_engagement_at, '0'),
  COALESCE(s.last_subscribed_at, '0'), -- this is excluded in WHERE above (s.last_subscribed_at < :thresholdDate)
  COALESCE(s.created_at, '0') -- this is also excluded by WHERE because created_at <= last_subscribed_at
) > :thresholdDate
```

I'll leave this question here for @costasovo, maybe, as it seems it would significantly simplify the code and probably also be more performant (no need for a temporary table, no need to join all again in the query below, simpler condition).

Reply:
Excellent point. 
I didn't look at the last query with the engagement values and it does look like it can be simplified. I also prefer that @costasovo takes a look in case we are overseeing something.

Could we move forward and review the query next week or do you think we need to wait? The deactivation criteria are very restrictive at the moment.

Reply:
> Could we move forward and review the query next week or do you think we need to wait?

Definitely. I know it's unrelated, I just noticed that...

Reply:
@brezocordero, @JanJakes Sorry for the late answer! I think your observations are correct. We changed the code couple of times, and now it looks like we could include those conditions into the temporary table query.
I looked into the file's history, and here is the commit [that added the temporary tables](https://github.com/mailpoet/mailpoet/pull/2401/files#diff-5ba4c0669779ae299f0e1ae5a8cba775c10e69d2d31792f1c94569e03145b37bR104-R124). At that time, the second query contained another join. Hence, it seemed reasonable to have it separated. Later [we changed the second query](https://github.com/mailpoet/mailpoet/commit/616b8868fe3e4b89c9628980431dd84ebb6f8b5a#diff-5ba4c0669779ae299f0e1ae5a8cba775c10e69d2d31792f1c94569e03145b37bR105-R114) to work with `last_engagement_at`, so after this change, we probably no longer need the separate query for the ids.

But looking at this PR, I found something I'm not sure is the correct change. To take also machine opens into account for inactive detection, [we started to treat machine opens as user engagement](https://github.com/mailpoet/mailpoet/pull/4004/commits/a59bfdbf427562fd01bcf6f2844e027f9b516718) and we update subscribers' `last_egagement_at` also for machine opens. This may have some side effects e.g. subscribers with machine-opens only won't receive re-engagement emails. I'm not sure if that is expected/ok behavior. Maybe @NeosinneR could confirm it is ok on not? Another approach we could take here is reintroducing the join on `mailpoet_statistics_opens` in the second query, which selects ids for deactivation (somewhat revert [this change](https://github.com/mailpoet/mailpoet/commit/616b8868fe3e4b89c9628980431dd84ebb6f8b5a#diff-5ba4c0669779ae299f0e1ae5a8cba775c10e69d2d31792f1c94569e03145b37bR105-R114)).

Reply:
I discussed the change @NeosinneR on slack and the change ⬆️  is correct.

Reply:
>  so after this change, we probably no longer need the separate query for the ids.

Thanks for confirming, I am working on another change that will touch the query and I will include the simplification.